### PR TITLE
Simplify preferences handling some more

### PR DESCRIPTION
**Pull request by Pentarctagon**

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
The purpose of the thing you're removing in this PR appears to be a way to allow packagers to set different default values than the normal Wesnoth defaults for some preferences. I don't know how useful that actually is, but it's not "an option for where the preferences file can be".

**Comment by Pentarctagon** [comment]
It's an option for where part of the preferences file can be, was my interpretation of it.

**Comment by Pentarctagon** [comment]
Looks like this was originally added by @lipk in 9ec70dabc502e9cc7e00b1dfd4e1cc4149c7dda0. No idea if he's still around, but if so then he can answer what it was for.

**Comment by soliton-** [comment]
The patch description for reference:

> Sometimes it's desired that packagers can distribute the game with custom settings. This patch enables an easy way to do that by adding the possibility to specify system-wide preferences in data/default-preferences.
> 
> Anything can be set in default-preferences which can be set the user preferences file and user preferences always override settings from default-preferences.

Further comments:
> lipk: Ivanovic, a concrete use case of this are packages which rename the binaries (such as my wesnoth-dev package for openSUSE or valdar's wesnoth-devel for Arch). There's currently no way to let wesnoth know if wesnothd's name was changed (when setting up a LAN game) without manually editing the user preferences file.

> ivanovic: Where would the system wide preferences have to reside? Basically I see two usecases for this:
>
> 1) Setting up the wesnoth client so that it can find the local wesnoth server or connect to some different server as default. This would mean some "wesnothrc" file.
> 
> 2) Define some server settings as default (port, sockets to use, ...).
> 
> Somehow those things should probably be defined somewhere in a file (or files) in /etc/ and things should be well integrated right in the build system(s) to also allow "normal users" (not only package managers who can bundle everything!) the use of this feature. For those cases this really sounds good and sane to me.

> boucman: ok, this patch works by adding a file in data/ that contains default values that can be overridden by the user
>
> it's a good start but not what Ivanovic wants...
>
> it would be nice if the patch would allow to specify a &amp;quot;system config location&amp;quot; at compile time (a scons/cmake parameter) and would also install the file at that location
>
> would that be fesible ?

> lipk: Here's the newest (and hopefully final) version of the patch.
> As requested by Ivanovic, packagers can set the location of the preferences file in compile time, and the binary won't try to load default preferences if a file wasn't specified in compile time.

> boucman: thx committed

**Comment by Pentarctagon** [comment]
I suppose that's a valid use case, if annoying. Where/how did you find the patch description?

**Comment by CelticMinstrel** [comment]
That use-case could also be solved with WML preprocessing, I think. Essentially, the preferences file would be a fixed file that _never_ changes. All it does is include a few other files. Then the non-syncable preferences are saved in one file, the syncable ones in another, and the default ones in a third, and the fixed preferences file just includes all three of them.

It probably isn't trivial to get that to work, especially since the preferences lie outside the virtual WML filesystem, but I think it should be possible.

**Comment by Pentarctagon** [comment]
If that's the only use case, it could just be implemented as a WESNOTHD_EXE_NAME define when building.

**Comment by Pentarctagon** [comment]
It'd also be pretty easy to have it be supported for LAN via the provided `lan_server.cfg`.

File: src/game_launcher.cpp [hunk 1]
```diff
@@ -820,23 +821,29 @@ bool game_launcher::goto_editor()
...
+	if (!filesystem::file_exists(path)) {
 		// copy file if it isn't created yet
-		filesystem::write_file(config, filesystem::read_file(filesystem::get_wml_location("lan_server.cfg")));
+		filesystem::write_file(path, filesystem::read_file(filesystem::get_wml_location("lan_server.cfg")));
+	}
+
+	config lan;
+	read(lan, path);
+
+	std::string wesnothd_program = lan["exe"].str("");
```
Comment (CelticMinstrel) [inline src/game_launcher.cpp:833]: I think it might be better to call this something like "binary" or "command".

Is making it be a shell script supported? If so, "command" is probably better; if not, I'd lean more towards "binary". It should also be made clear whether it can be a shell _command_ (as in, whether this string will itself be parsed as a command). From the code I _think_ it looks like it's not? I'm not quite sure, though.

Comment (Pentarctagon) [reply src/game_launcher.cpp]: Renamed to binary - a shell command would fail the file exists check.

Comment (CelticMinstrel) [reply src/game_launcher.cpp]: A shell _command_ would, but not a shell _script_.

Comment (Pentarctagon) [reply src/game_launcher.cpp]: Ah, then yeah, a shell script would probably work as well. Changed to command then.

Comment (CelticMinstrel) [reply src/game_launcher.cpp]: The distinction (that it must be a valid file path and is not interpreted by the shell) can just be put in documentation somewhere.

Comment (Pentarctagon) [reply src/game_launcher.cpp]: Added to the help output of wesnothd and the wesnothd man page.

**Comment by soliton-** [comment]
The patch description is from the GNA patch export somewhere on files.wesnoth.org.

It's a bit odd that the server config now contains an attribute specifying how to start the server but I guess it makes sense for the case that wesnoth invokes the server. Just need to keep that in mind for future server config additions.

(I'm currently on the road so can't review that well.)

File: doc/man/wesnothd.6 [hunk 2]
```diff
@@ -49,6 +49,7 @@ on what commands the server accepts via the wesnoth client (/query ...) or the f
 tells wesnothd where to find the config file to use. See the section
 .B SERVER CONFIG
 below for the syntax. You can reload the config with sending SIGHUP to the server process.
+When used as a LAN server, the 'command' attribute can be used to where to specify either this executable or a script file to run this executable.
```
Comment (CelticMinstrel) [inline doc/man/wesnothd.6:52]: This confuses me. What does it have to do with the changes made in this PR?

File: src/server/wesnothd/server.cpp [hunk 3]
```diff
@@ -3051,6 +3051,9 @@ int main(int argc, char** argv)
 			std::cout << "usage: " << argv[0]
 					  << " [-dvV] [-c path] [-m n] [-p port] [-t n]\n"
 					  << "  -c, --config <path>        Tells wesnothd where to find the config file to use.\n"
+					  << "                             When used as a LAN server, the 'command' attribute can\n"
```
Comment (CelticMinstrel) [inline src/server/wesnothd/server.cpp:3054]: Again… how is this related to this PR at all?

File: doc/man/wesnothd.6 (hunk 2 above)
Comment (Pentarctagon) [reply doc/man/wesnothd.6:52]: That's part of what the PR does?

Comment (CelticMinstrel) [reply doc/man/wesnothd.6:52]: What? There isn't anything in this PR that changes the server though?

Comment (Pentarctagon) [reply doc/man/wesnothd.6:52]: It adds an attribute that can be present in the server config that gets passed to wesnothd when the wesnoth client starts it as a LAN server. I don't mind putting that documentation somewhere else, but I couldn't think of any better place.

Comment (CelticMinstrel) [reply doc/man/wesnothd.6:52]: Hmm. It seems I misunderstood something. The place you added `command` was not the place I thought you added it. I was thinking of a tag that is probably part of the preferences file, though I don't see it in my own preferences file so it's hard to check…

I now understand what soliton was talking about. It does seem very odd for this config file to contain the `command` key.

Comment (Pentarctagon) [reply doc/man/wesnothd.6:52]: The preferences already has an attribute for that, so it wouldn't make sense to add a second attribute for it. The other alternative of course still being to add a define specifically for the wesnothd location.

Comment (CelticMinstrel) [reply doc/man/wesnothd.6:52]: I don't understand what you're saying… the preferences already has the thing you added? So why did you add it if it already exists? I'm _really_ confused now.

Comment (Pentarctagon) [reply doc/man/wesnothd.6:52]: The current functionality allows specifying a "system preferences file" that can be used to provide default values for preferences.

---

The use case for this functionality was apparently to:
1) Locate the wesnothd executable when it's been given a different name than the Wesnoth client expects when packaging for Linux distros. This is set using the `mp_server_program_name` preference attribute.
2) Specify some other server settings such as the port to use - this was apparently never actually implemented.

---

This PR removes that "system preferences file" handling using the `mp_server_program_name` preferences attribute and replaces it with the `command` attribute that Linux packagers can append to [lan_server.cfg](https://github.com/wesnoth/wesnoth/blob/master/data/lan_server.cfg).

---

The result being that for any packager who's using this (and so far only the Arch packager has replied to the email I sent to the packager's ML to say he doesn't use it for Arch), instead of creating a "system preferences file" containing just `mp_server_program_name=/path/to/wesnothd` and also compiling Wesnoth with the `DEFAULT_PREFS_PATH=/path/to/system/preferences` they would instead just append `command=/path/to/wesnothd` to `data/lan_server.cfg`.

Comment (CelticMinstrel) [reply doc/man/wesnothd.6:52]: It is definitely a weird place to put that information. Preferences or GameConfigWML seem like a better option to me. Not sure how viable the latter would actually be for packagers, mind you.

Comment (Pentarctagon) [reply doc/man/wesnothd.6:52]: It's not that weird - it's putting the location for where to find wesnothd for LAN in the file with other LAN configuration settings. IMO that makes a lot more sense than having to setup a separate preferences file and then build Wesnoth with a compile option pointing to that file.

Comment (CelticMinstrel) [reply doc/man/wesnothd.6:52]: Except for the fact that that file is intended for settings to configure the server once it has started, not settings that the client uses to find the server.

Comment (Pentarctagon) [reply doc/man/wesnothd.6:52]: I don't think that's a reason it can't be used for this though.

**Comment by Pentarctagon** [comment]
@rhonda is `DEFAULT_PREFS_FILE` something that you use?

**Comment by Pentarctagon** [comment]
To add it here: rhonda replied from IRC and said it's not used for Debian either.

**Comment by soliton-** [comment]
Note that debian instead patches the wesnoth sources: https://salsa.debian.org/games-team/wesnoth/-/blob/38650cf6ba513502ce79d263cdbbb850c112f219/debian/patches/03wesnothd-name

That is not a good thing.

On a quick look it seems arch does not package multiple wesnoth versions in parallel.

**Comment by soliton-** [comment]
So essentially this PR is replacing generic default preferences logic with a single oddly placed attribute for the wesnothd command. IMO that is not an improvement. Is there some other reason to do this besides "simplification"?

**Comment by Pentarctagon** [comment]
I can't seem to view that Debian page - it just gives me a 500 error.

---

It's replacing a feature that is unused with a simpler way to achieve the same result for its use case if someone ends up wanting it.

**Comment by soliton-** [comment]
I think if we want a simpler way to find wesnothd when it and wesnoth don't have the usual names it'd be much easier to expect that both where renamed in the same way. So if wesnoth is called wesnoth-1.18 then we look for wesnothd-1.18. That would require the packager to do nothing at all and I imagine that would cover pretty much all cases.

I do not agree that the default preferences logic is complicated at all. It seems perfectly trivial and is a general solution for a packager to provide default settings for anything. They can specify their own MP or campaign server for example. Even if it's not particularly used right now it seems perfectly useful and simple to me. If we want to remove that feature IMO that is a separate issue and should not be done in a commit called "Simplify preferences handling".

**Comment by rhonda** [comment]
> I can't seem to view that Debian page - it just gives me a 500 error.
> 
> It's replacing a feature that is unused with a simpler way to achieve the same result for its use case if someone ends up wanting it.

Yep, salsa is currently undergoing DoS attacks, sorry about that. The patch does change the hardcoded "wesnothd" in src/game_launcher.cpp and src/preferences/display.cpp to the versioned wesnothd-1.18 (e.g.) to be able to install different releases side-by-side (for people that do system upgrades and want to continue their old savegames without having to start over new).

I'm unsure currently if I did this at a point the flag wasn't there yet, or whether I simply haven't found the flag to use, or this patch looked more direct to what I wanted.  The switch wouldn't e.g. change the LOG message neither (so you potentially can't differentiate from the log which one was started, but that might be a very smallish thing).

The feature itself is useful for us - though as mentioned we don't use it but patch the source. :)
